air/src/execution_step/lambda_applier/applier.rs

```diff
@@ -104,13 +99,43 @@ fn select_by_path_from_stream<'value>(
     Ok(select_result)
 }
 
+fn select_by_path_from_canon_map_stream<'value>(
+    stream: impl ExactSizeIterator<Item = (&'value JValue, RcSecurityTetraplet)> + 'value,
+    lambda: &NonEmpty<ValueAccessor<'_>>,
+    exec_ctx: &ExecutionCtx<'_>,
+) -> ExecutionResult<MapLensResult<'value>> {
+    let stream_size = stream.len();
+    let (idx, body) = exctract_idx(lambda, exec_ctx)?;
+
+    let (value, tetraplet) = lambda_to_execution_error!(stream
+        .peekable()
+        .nth(idx)
+        .ok_or(LambdaError::CanonStreamNotHaveEnoughValues { stream_size, idx }))?;
+
+    // csm.$.key.[0] case
+    let select_result = if body.is_empty() {
+        let result = Cow::Borrowed(value);
+        MapLensResult::from_cow(result, tetraplet)
+    } else {
+        // csm.$.key.[0].attribute case
+        let result = select_by_path_from_scalar(value, body.iter(), exec_ctx)?;
+
+        let json_path_suffix = body.iter().fold("".to_string(), |acc, va| acc + "." + &va.to_string());
+        let prefix_with_path = true;
+        let tetraplet = update_tetraplet_with_path(&tetraplet, &json_path_suffix, prefix_with_path);
+
+        MapLensResult::from_cow(result, tetraplet)
```

Comment:
`...::from_cow` call is common for both branches.  Extracting common code reduces code complexity.

Reply:
I've been tricked by the shadowing of the `tetraplet` variable.  Such shadowing makes code harder to read.